### PR TITLE
FIX: Fixes unexpected behaviour for Llava / LLama & AWQ Fused modules + revert #30070 at the same time

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

Fixes a silent behaviour introduced by a recent PR, passing a `None` attention mask results in unexpected behaviour for awq fused modules. The fix is simply to force set a dummy `_attn_implementation` on the config objects of the modules that contain fused modules

I can confirm the failing slow tests now pass with these changes

cc @ArthurZucker @fxmarty

### CODE REVIEW

File: src/transformers/integrations/awq.py [hunk 1]
```diff
@@ -241,7 +243,27 @@ def fuse_awq_modules(model, quantization_config):
...
+        fused_attention_parent_modules = {
+            fused_attention_module.split(".")[0] for fused_attention_module in fused_attention_modules
+        }
+        for module_name, module in model.named_modules():
+            if any(
+                module_name in fused_attention_parent_module
+                for fused_attention_parent_module in fused_attention_parent_modules
+            ):
+                if hasattr(module, "config") and hasattr(module.config, "_attn_implementation"):
+                    module.config._attn_implementation = "custom"
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/transformers/integrations/awq.py:266]: because some models such as llava have other components such as the CLIP vision backbone, we want to keep them untouched and use the passed attn_implementation

File: src/transformers/modeling_attn_mask_utils.py [hunk 2]
```diff
@@ -351,6 +404,14 @@ def _prepare_4d_causal_attention_mask_for_sdpa(
...
             # may be wrong. We will set `is_causal=False` in SDPA and rely on Transformers attention_mask instead, hence not setting it to None here.
             # Reference: https://github.com/pytorch/pytorch/issues/108108
+=======
+    ignore_causal_mask = AttentionMaskConverter._ignore_causal_mask_sdpa(
+        attention_mask=attention_mask,
+        inputs_embeds=inputs_embeds,
+        past_key_values_length=past_key_values_length,
+        sliding_window=sliding_window,
+    )
+>>>>>>> parent of acab997bef... Revert "Re-enable SDPA's FA2 path (#30070)" (#30314)
```
Comment (fxmarty) [inline src/transformers/modeling_attn_mask_utils.py:414]: beware

File: src/transformers/modeling_attn_mask_utils.py [hunk 3]
```diff
@@ -316,6 +368,7 @@ def _prepare_4d_causal_attention_mask_for_sdpa(
         or (hasattr(torch, "_dynamo") and torch._dynamo.is_compiling())
     )
 
+<<<<<<< HEAD
```
Comment (fxmarty) [inline src/transformers/modeling_attn_mask_utils.py:371]: same

Comment (younesbelkada) [reply src/transformers/modeling_attn_mask_utils.py]: nice catch!

File: src/transformers/integrations/awq.py [hunk 4]
```diff
@@ -241,7 +243,23 @@ def fuse_awq_modules(model, quantization_config):
...
+        attention_has_been_fused = _fuse_awq_attention_layers(
+            model, module, modules_to_fuse, name, QuantAttentionFused
+        )
+
+        if attention_has_been_fused:
+            fused_attention_modules.append(name.split(".")[0])
+
+    # For AWQ fused + Llama we need to set `config._attn_implementation` = "custom" to avoid unexpected behavior and pass
+    # `None` attention mask to the fused attention modules as now the attention mask is dropped by our models and dealt
+    # by the `AttentionMaskConverter` module.
```
Comment (ArthurZucker) [inline src/transformers/integrations/awq.py:255]:
```suggestion
    # For AWQ fused + Llama we need to set `config._attn_implementation` = "custom" to avoid unexpected behaviors. We loop over the layers to make sure the vision/text config are
    # modified only if some of their modules were fused
```

**Review by ArthurZucker — APPROVED** [review]
Let's make sure to run the slow Llama tests at least before merging to iterate fast if they fail

File: src/transformers/integrations/awq.py (hunk 4 above)
Comment (ArthurZucker) [reply src/transformers/integrations/awq.py:255]: nit but more understandable

**Review by fxmarty — APPROVED** [review]
Please apply the above fix and LGTM

If you can run
```python
from transformers import WhisperForCausalLM, WhisperForConditionalGeneration, WhisperProcessor
import torch
from datasets import load_dataset

processor = WhisperProcessor.from_pretrained("openai/whisper-large-v2")
model = WhisperForConditionalGeneration.from_pretrained("openai/whisper-large-v2")

assistant_model = WhisperForCausalLM.from_pretrained("distil-whisper/distil-large-v2")

ds = load_dataset("hf-internal-testing/librispeech_asr_dummy", "clean", split="validation")
sample = ds[0]["audio"]
input_features = processor(
    sample["array"], sampling_rate=sample["sampling_rate"], return_tensors="pt"
).input_features

predicted_ids = model.generate(input_features, assistant_model=assistant_model)

# decode token ids to text
transcription = processor.batch_decode(predicted_ids, skip_special_tokens=True)[0]
print(transcription)
```

& llava AWQ test & llama slow test to be sure as well

**Comment by younesbelkada** [comment]
Thanks @fxmarty @ArthurZucker ! just tested everything seems to all pass !
